File: app/src/main/java/app/grapheneos/pdfviewer/GestureHelper.java
package app.grapheneos.pdfviewer;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.concurrent.atomic.AtomicBoolean;

/*
    The GestureHelper present a simple gesture api for the PdfViewer
*/

class GestureHelper {
    public interface GestureListener {
        boolean onTapUp();
        boolean onFling(@Nullable MotionEvent e1, @NonNull MotionEvent e2, float velocityX, float velocityY);
        void onZoom(float scaleFactor, float focusX, float focusY);
        void onZoomEnd();
    }

    @SuppressLint("ClickableViewAccessibility")
    static void attach(Context context, View gestureView, GestureListener listener) {

        AtomicBoolean wasScaling = new AtomicBoolean(false);

        final ScaleGestureDetector scaleDetector = new ScaleGestureDetector(context,
                new ScaleGestureDetector.SimpleOnScaleGestureListener() {
                    @Override
                    public boolean onScale(@NonNull ScaleGestureDetector detector) {
                        listener.onZoom(detector.getScaleFactor(), detector.getFocusX(),
                                detector.getFocusY());

                        return true;
                    }

                    @Override
                    public void onScaleEnd(@NonNull ScaleGestureDetector detector) {
                        listener.onZoomEnd();
                    }
                });

        final GestureDetector detector = new GestureDetector(context,
                new GestureDetector.SimpleOnGestureListener() {
                    @Override
                    public boolean onSingleTapUp(@NonNull MotionEvent motionEvent) {
                        return listener.onTapUp();
                    }

                    @Override
                    public boolean onFling(@Nullable MotionEvent e1, @NonNull MotionEvent e2,
                                           float velocityX, float velocityY) {
                        if (wasScaling.get()) {
                            return false;
                        }

                        return listener.onFling(e1, e2, velocityX, velocityY);
                    }
                });

        gestureView.setOnTouchListener((view, motionEvent) -> {
            int action = motionEvent.getActionMasked();

            if (action == MotionEvent.ACTION_DOWN) {
                wasScaling.set(false);
            }

            detector.onTouchEvent(motionEvent);
            scaleDetector.onTouchEvent(motionEvent);

            // Check after scaleDetector processes the event
            if (scaleDetector.isInProgress()) {
                wasScaling.set(true);
            }

            return false;
        });
    }

}

File: app/src/main/java/app/grapheneos/pdfviewer/GestureHelper.kt
package app.grapheneos.pdfviewer

import android.annotation.SuppressLint
import android.content.Context
import android.view.InputDevice
import android.view.GestureDetector
import android.view.KeyEvent
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.view.View
import kotlin.math.roundToInt

/*
 * The GestureHelper present a simple gesture api for the PdfViewer
 */

object GestureHelper {
    interface GestureListener {
        fun onTapUp(): Boolean
        fun onFling(e1: MotionEvent?, e2: MotionEvent, velocityX: Float, velocityY: Float): Boolean
        fun onZoom(scaleFactor: Float, focusX: Float, focusY: Float)
        fun onCtrlMouseWheelZoom(zoomIn: Boolean, focusX: Float, focusY: Float)
        fun onZoomEnd()
    }

    @SuppressLint("ClickableViewAccessibility")
    fun attach(context: Context, gestureView: View, listener: GestureListener) {
        var wasScaling = false
        var wheelTickRemainder = 0f

        val scaleDetector = ScaleGestureDetector(
            context,
            object : ScaleGestureDetector.SimpleOnScaleGestureListener() {
                override fun onScale(detector: ScaleGestureDetector): Boolean {
                    listener.onZoom(detector.scaleFactor, detector.focusX, detector.focusY)
                    return true
                }

                override fun onScaleEnd(detector: ScaleGestureDetector) {
                    listener.onZoomEnd()
                }
            })

        val detector = GestureDetector(
            context,
            object : GestureDetector.SimpleOnGestureListener() {
                override fun onSingleTapUp(e: MotionEvent): Boolean {
                    return listener.onTapUp()
                }

                override fun onFling(
                    e1: MotionEvent?,
                    e2: MotionEvent,
                    velocityX: Float,
                    velocityY: Float
                ): Boolean {
                    if (wasScaling) return false
                    return listener.onFling(e1, e2, velocityX, velocityY)
                }
            })

        gestureView.setOnTouchListener { _, event ->
            if (event.actionMasked == MotionEvent.ACTION_DOWN) {
                wasScaling = false
            }

            detector.onTouchEvent(event)
            scaleDetector.onTouchEvent(event)

            if (scaleDetector.isInProgress) {
                wasScaling = true
            }

            false
        }

        gestureView.setOnGenericMotionListener { _, event ->
            if (!isCtrlPhysicalMouseWheelEvent(event)) {
                return@setOnGenericMotionListener false
            }

            wheelTickRemainder += event.getAxisValue(MotionEvent.AXIS_VSCROLL)
            val wholeTickDelta = wheelTickRemainder.roundToInt()
            wheelTickRemainder -= wholeTickDelta
            if (wholeTickDelta != 0) {
                listener.onCtrlMouseWheelZoom(
                    zoomIn = wholeTickDelta > 0,
                    focusX = event.x,
                    focusY = event.y
                )
            }

            true
        }
    }

    private fun isCtrlPhysicalMouseWheelEvent(event: MotionEvent): Boolean {
        return event.actionMasked == MotionEvent.ACTION_SCROLL &&
                event.pointerCount > 0 &&
                (event.metaState and KeyEvent.META_CTRL_ON) != 0 &&
                (event.source and InputDevice.SOURCE_MOUSE) == InputDevice.SOURCE_MOUSE &&
                event.getToolType(0) == MotionEvent.TOOL_TYPE_MOUSE &&
                event.getAxisValue(MotionEvent.AXIS_VSCROLL) != 0f
    }
}
